### PR TITLE
Add Zafu/Text/Pretty pretty-printer module (issue #51)

### LINKED ISSUE #51: Add a pretty printer
Read the API and code of this: https://github.com/typelevel/paiges

Design a Zafu/Text/Pretty package Doc type with basically the same API as paiges.Doc. Add a `struct Document[a]` typeclass-pattern code that exposes: `def to_Doc[a](docu: Document[a], item: a) -> Doc` and the `by` combinator similar to `Hash/Eq`.

Make a list of all the public exports of the code. The API should support the most efficient possible implementation (which is an excellent balance of  memory and runtime efficiency).

### PR DESCRIPTION
**Pull request by johnynek**
Implemented `src/Zafu/Text/Pretty.bosatsu` with the new pretty-printing API and tests from the merged design scope.

What was added:
- New `Doc` core with internal `VDoc` node representation and exported combinators/constants (`empty`, `line`, `line_or_space`, `grouped`, `flatten`, `render*`, etc.).
- Rendering pipeline with width-sensitive branch selection (`Union` + `fits`), wide rendering path, trim rendering, and stream-style render outputs.
- Structural helpers and collection combinators (`append*`, `hang`, `indent`, `bracket_by`, `fill`, `intercalate`, `cat`, `spread`, `stack`, `tabulate`, `split`, `paragraph`, etc.).
- `Document[a]` typeclass-pattern API with `document_from_fn`, `to_Doc`, `by`, plus required predefined instances: `document_String`, `document_Char`, `document_Unit`, `document_Bool`, `document_Int`, `document_Float64`, `document_List`, `document_Option`.
- Numeric formatting entry points `int` and `float`, including explicit handling for `NaN`, infinities, and `-0.0`.
- In-module tests covering representative rendering behavior, trim behavior, `Document` behavior, numeric edge cases, and deep-doc stack-safety regression checks.

Validation run:
- `./bosatsu lib check` passed
- `./bosatsu lib test` passed
- `scripts/test.sh` passed

Fixes #51

Implements design doc: [docs/design/51-add-a-pretty-printer.md](https://github.com/johnynek/zafu/blob/main/docs/design/51-add-a-pretty-printer.md)

Design source PR: https://github.com/johnynek/zafu/pull/54

### CODE REVIEW

File: src/Zafu/Text/Pretty.bosatsu [hunk 1]
```diff
@@ -0,0 +1,1069 @@
...
+  loop(force_v_limit, doc)
+
+def traversal_fuel(doc: Doc, width_hint: Int) -> Int:
+  size = doc_size_v(view(doc))
+  size.mul(max_Int(width_hint, 0).add(32)).add(1024)
+
+def string_is_empty(value: String) -> Bool:
+  cmp_String(value, "") matches EQ
+
+space_char_opt: Option[Char] = int_to_Char(32)
```
Comment (johnynek) [inline src/Zafu/Text/Pretty.bosatsu]: why not `space_char = .' '` we have literal characters

**Review by johnynek — COMMENTED** [review]
every exported function needs a detailed comment directly above it to explain the inputs and the outputs so people only reading the doc output can understand how to use the library.

File: src/Zafu/Text/Pretty.bosatsu [hunk 2]
```diff
@@ -0,0 +1,1069 @@
...
+  Document(to_doc_fn) = docu
+  to_doc_fn(value)
+
+def by(docu: Document[b], project: a -> b) -> Document[a]:
+  document_from_fn(value -> to_Doc(docu, project(value)))
+
+document_String: Document[String] = document_from_fn(text)
+document_Char: Document[Char] = document_from_fn(char)
+document_Unit: Document[Unit] = document_from_fn(_ -> text_raw("()"))
+document_Bool: Document[Bool] = document_from_fn(value -> text_raw("True") if value else text_raw("False"))
```
[the remark this answers is not in the document]
Comment (johnynek) [reply src/Zafu/Text/Pretty.bosatsu]: Hoisted reusable bool docs to top-level constants (`true_Doc`, `false_Doc`) and switched `document_Bool` to reuse them so we avoid per-call allocation.

**Review by johnynek — COMMENTED** [review]
fix the fuel + force_v + stack pattern by using `recur (fuel, stack)` to minimize fuel consumption.

File: src/Zafu/Text/Pretty.bosatsu [hunk 3]
```diff
@@ -0,0 +1,1298 @@
...
+                else:
+                  False
+              case Line:
+                False
+              case Union(flattened, _):
+                loop(next_fuel, [flattened, *tail])
+              case LazyDoc(thunk):
+                loop(next_fuel, [get_Lazy(thunk), *tail])
+
+  loop(doc_size_v(view(doc)).mul(8).add(64), [view(doc)])
```
[the remark this answers is not in the document]
Comment (johnynek) [reply src/Zafu/Text/Pretty.bosatsu]: Applied. I removed repeated `view(doc)` calls by caching `root = view(doc)` and reusing it in both `is_empty` and `max_width` call sites.

**Comment by johnynek** [comment]
Implemented feedback from issue comment 4034892295 by removing the `VDoc` wrapper and using `Doc` as the enum throughout `src/Zafu/Text/Pretty.bosatsu`. Internal traversal/render helpers now operate directly on `Doc`, and the `view`/`mk` indirection has been removed from call sites. Validation completed with `./bosatsu --fetch lib check` and required pre-push `scripts/test.sh` (both passing).
